core/services/relay/evm/chain_reader.go

```diff
@@ -269,59 +258,117 @@ func (cr *chainReader) addEvent(contractName, eventName string, a abi.ABI, chain
-			eb.WithTopic(genericTopicName, topic, uint64(topicIndex))
-		}
+			// Encoder defs codec won't be used for encoding, but for storing caller filtering params which won't be hashed.
+			topicTypeID := eventName + "." + genericTopicName
+			err := cr.addEncoderDef(contractName, topicTypeID, abi.Arguments{{Type: topic.Type}}, nil, inputModifications)
+			if err != nil {
+				return nil, nil, nil, err
+			}
 
-		cr.bindings.AddReader(contractName, genericTopicName, eb)
-	}
+			topicTypeID = codec.WrapItemType(contractName, topicTypeID, true)
+			topicsTypes[topicTypeID], topicsModifiers[topicTypeID], err = cr.getEventItemTypeAndModifier(topicTypeID, inputModifications)
+			if err != nil {
+				return nil, nil, nil, err
+			}
 
-	// add data word readBindings for QueryKey
-	for genericDataWordName := range eb.GetDataWords() {
-		cr.bindings.AddReader(contractName, genericDataWordName, eb)
+			topicsDetails[genericTopicName] = read.TopicDetail{Argument: topic, Index: uint64(topicIndex + 1)}
+		}
 	}
+	return topicsDetails, topicsTypes, topicsModifiers, nil
 }
 
-// getEventInput returns codec entry for expected incoming event params and the modifier to be applied to the params.
-func (cr *chainReader) getEventInput(def types.ChainReaderDefinition, contractName, eventName string) (
-	types.CodecEntry, commoncodec.Modifier, error) {
-	inputInfo := cr.parsed.EncoderDefs[WrapItemType(contractName, eventName, true)]
+// initDWQuerying registers codec types for evm data words to be used for typing value comparator QueryKey filters.
+func (cr *chainReader) initDWQuerying(contractName, eventName string, eventDWs map[string]read.DataWordDetail, dWDefs map[string]types.DataWordDef) (map[string]read.DataWordDetail, map[string]types.CodecEntry, error) {
+	dwsCodecTypeInfo := make(map[string]types.CodecEntry)
+	dWsDetail := make(map[string]read.DataWordDetail)
+
+	for genericName, cfgDWInfo := range dWDefs {
+		foundDW := false
+		for _, dWDetail := range eventDWs {
+			// err if index is same but name is diff
+			if dWDetail.Index != cfgDWInfo.Index {
+				if dWDetail.Name == cfgDWInfo.OnChainName {
+					return nil, nil, fmt.Errorf("failed to find data word with index %d for event: %q data word: %q", dWDetail.Index, eventName, genericName)
+				}
+				continue
+			}
+
+			foundDW = true
+
+			dwTypeID := eventName + "." + genericName
+			if err := cr.addEncoderDef(contractName, dwTypeID, abi.Arguments{abi.Argument{Type: dWDetail.Type}}, nil, nil); err != nil {
+				return nil, nil, fmt.Errorf("%w: failed to init codec for data word %s on index %d querying for event: %q", err, genericName, dWDetail.Index, eventName)
+			}
+
+			dwCodecTypeID := codec.WrapItemType(contractName, dwTypeID, true)
+			dwsCodecTypeInfo[dwCodecTypeID] = cr.parsed.EncoderDefs[dwCodecTypeID]
 
-	// TODO can this be simplified? Isn't this same as inputInfo.Modifier()? BCI-3909
-	inMod, err := def.InputModifications.ToModifier(codec.DecoderHooks...)
+			dWsDetail[genericName] = dWDetail
```

Comment:
Just realised that this [PR](https://github.com/smartcontractkit/chainlink/pull/13992) ran over this [PR](https://github.com/smartcontractkit/chainlink/pull/14096), so this TODO is getting re-resolved